[feat] experimental MEVO layer

min-xu-ai opened the pull request:
## What does this PR do?

This introduces a custom NN layer called MemoryEfficientVocabOutput (MEVO). This layer includes a linear output projection and a crossentropy loss (softmax + NLL loss). The main goal for this custom layer is to support very large scale training with large batch size, long sequence length and large vocabulary size.

The techniques uses in this layer are the following:

1. custom fused, checkpointed, tiled functions that avoiding materializing a HUGE output matrix from the output projection.
2. custom backward to avoid multiple copies of gradient buffers.
3. custom FSDP wrapping to shard the parameter and gradients for the projection.

A benchmark and 2 unit tests are added to test this layer. The layer is compared with a baseline implementation for both correctness, memory and speed. The speed is largely reduced due to complex python pytorch operator sequences. In the future, we will address the speed issues. At this moment, due to reduced memory, we can increase the batch size for the overall model training and gain moderate speedup overall.

The associated FSDP change is already merged with #836. It might be useful to refer to that commit while reviewing.

Anyone is welcome to review and provide constructive feedback. Thanks!

This is experimental and once it is more validated and more ready for general consumption, I will add more extensive user doc.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A


## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

min-xu-ai commented:
This is ready for review. Please take a look whoever has the time. No pressure.

It is still experimental since I have yet to run end to end tests with loss curves. Unit tests do ensure numerical comparisons though. Also, speed is not great. More work is needed in the future, likely with Triton.

I will have a separate doc with detailed performance and memory analysis.

min-xu-ai commented on fairscale/experimental/nn/ssd_offload.py:
```diff
@@ -166,9 +166,9 @@ def copy_into_tensor(self, tensor: torch.Tensor) -> None:
 
         If the tensor is in memory, this function copies the data
-        into the passed in tensor. Otherwise, it reads from file into tensor, 
+        into the passed in tensor. Otherwise, it reads from file into tensor,
```
these whilespace change is added by pre-commit hook.

min-xu-ai commented on fairscale/nn/data_parallel/fully_sharded_data_parallel.py:
```diff
@@ -1412,7 +1412,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 param = param._linked_param
 
-        assert param.grad is not None
+        assert param.grad is not None, param.shape
```
this single line is for better debugging output.

blefaudeux commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,480 @@
+        ctx.w_split_size = w_split_size
+        ctx.args = {}
+        assert split_dim == 0
+        with torch.no_grad():
```
this is to save some memory/speed, because autograd is wrong there ? curious as of why this is needed basically

min-xu-ai replied:
This is for saving memory with checkpointing. I will have an extensive doc that describes the algorithm in more details. Will add more comment here too.

blefaudeux commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,480 @@
+        i, w = ctx.saved_tensors
+        assert i.requires_grad
+        assert w.requires_grad
+        i = i.detach().requires_grad_(True)
```
the sequence here (detach, FW/BW with autograd, then no_grad again) is pretty confusing without much context. Could you add some explanations ? I'm a bit worried about future maintainability

min-xu-ai replied:
Yes, very true! I will add more comments in the code around this.

anj-s replied:
thanks for adding the comments! they definitely help but I think we need a doc to be honest :) this is pretty complex and context is pretty low with just the code.

blefaudeux commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,480 @@
+            mem = round(torch.cuda.max_memory_allocated() / 1024 / 1024)
+            print("DEBUG cur, peak", cur_mem, mem)
+        assert kwargs == {}
+        input, target = input
```
I don't get that, why not declare them directly as inputs and get some help from mypy ?

min-xu-ai replied:
thanks for the review, Ben!

if I recall this correctly, the prototype for `forward` in mypy is actually the Any, Any one. That's why I used that to make mypy happier. If you know a way that I can use separate args more directly here without too much mypy magic, that'd be great!

blefaudeux replied:
oh I typically ignore the mypy warning on this line ( #type: ignore), because I think that declaring inputs helps with readability (a lot of people will look at the interface, a lot more than the number of people who will look at the code in any case). The issue with the inputs goes with Pytorch not being typed in general I think, so this #type: ignore is a little ugly but I can live with it

min-xu-ai commented:
@blefaudeux, addressed your comments. Thanks for the review!

If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.

anj-s commented:
> @blefaudeux, addressed your comments. Thanks for the review!
> 
> If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.



> @blefaudeux, addressed your comments. Thanks for the review!
> 
> If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.

Taking a look now, would be great if you could wait till tomorrow morning. thanks!

min-xu-ai commented:
> Taking a look now, would be great if you could wait till tomorrow morning. thanks!

Thanks for taking a look! Of course!

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+            The shared weight.
+        tile_factor (int):
+            Unused. It is here to make kernel init easier with MEVO.
+        log (bool):
```
nit: I would say "use_log" instead of "log". makes it more clear that it is a bool.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+        input, target = _reshape_inputs(input, target)
+        x = super().forward(input, target)
+        x = F.nll_loss(x, target, reduction="sum")
+        return x
```
nit: early return in line 129?

min-xu-ai replied:
sure.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+
+
+class GetMaxFunction(torch.autograd.Function):
+    """Custom checkpointed function to get max-per-token from an input and a weight"""
```
nit: missing period at the end of this sentence.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+    if len(input.shape) == 3:
+        input = input.reshape(-1, input.shape[2])
+    if len(target.shape) == 2:
+        target = target.reshape(-1)
```
will this be a 1-D or 2D target?

min-xu-ai replied:
target is always 1D in this layer. We just reshape the input when the input is 2D. input is also always reshaped from 3d to 2d.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+    def get_max(i: torch.Tensor, w: torch.Tensor, fp: bool) -> torch.Tensor:
+        """ i: (split-of-tokens, d_model), w: (split-of-vocabs, d_model)"""
+        _m = torch.matmul(i, w.T)
+        if fp:
```
what does fp stand for?

min-xu-ai replied:
fp is full precision, sharing the same acronym with floating point. :-)

I will make those longer names.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+    """Custom checkpointed function to compute the target score."""
+
+    @staticmethod
+    def get_target_score(i: torch.Tensor, w: torch.Tensor, target: torch.Tensor, fp: bool) -> torch.Tensor:
```
can we rename i and w across this file to something that it represents? it might be easier to grok this code.

min-xu-ai replied:
Alright, I added an explanation of them at the first occurrence.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+class MemoryEfficientVocabOutput(nn.Module):  # AKA. MEVO
+    """ Fused fc + softmax + nll_loss in a tiled fashion.
+
+        This uses much less memory but is quite a bit slower.
```
just curious about why it is slower. do we know why?

min-xu-ai replied:
Yes, we know some of it is due to re-computation. Others are likely due to python looping. This is definitely an important future work.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
```
Why do we implement this? is it for testing convenience?

min-xu-ai replied:
correct. This is purely for testing and benchmarking.

min-xu-ai replied:
added a comment.

anj-s commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -0,0 +1,483 @@
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 16, reduction: str = "sum"):
+        super().__init__()
+        self.proj_weight = proj_weight
+        self.tf_in, self.tf_w = tile_factor, tile_factor
```
does tile_factor have to be the same? how does it impact memory savings?

min-xu-ai replied:
It doesn't need to be the same. I haven't extensively tuning it. Will add a TODO.

anj-s commented on tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py:
```diff
@@ -0,0 +1,160 @@
+            # Shared layers much be un-flatten.
+            self.l0 = FSDP(self.l0, flatten_parameters=False, mixed_precision=False, compute_dtype=torch.float16)
+            self.l1 = FSDP(self.l1, flatten_parameters=False, mixed_precision=False, compute_dtype=torch.float16)
+            self.l1.append_shared_param(self.l0.module.weight)
```
MEVO should also work for non-shared layers right? Curious what the reason was for having that support.

min-xu-ai replied:
The floor memory consumption won't be as good otherwise. If mevo is used without this sharding support, then its param and grad memory will stay unsharded (i.e. full) for the entire forward/backward pass. This reduces the memory usable by other layers. With this support, we shard them after the layer is done computing, leaving much more memory to other layers, like the attention layers.

anj-s commented:
nice PR @min-xu-ai ! Would love to see a doc when you get a chance. Feel free to merge since my comments are more questions with a few nits. Without a doc I am not sure I can comment about the algorithm anyway. Thanks for waiting for the merge.

min-xu-ai commented:
> nice PR @min-xu-ai ! Would love to see a doc when you get a chance. Feel free to merge since my comments are more questions with a few nits. Without a doc I am not sure I can comment about the algorithm anyway. Thanks for waiting for the merge.

Thank you so much for the review @anj-s! Yes, the doc is coming too. The comments and suggestions are great. Will address them!

blefaudeux commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -165,6 +161,9 @@ def forward(  # type: ignore
         ctx.w_split_size = w_split_size
         ctx.args = {}
         assert split_dim == 0
+        # During forward, we use ``no_grad'' to avoid saving the activations.
```
<3

blefaudeux commented on fairscale/experimental/nn/mevo.py:
```diff
@@ -184,15 +183,21 @@ def backward(ctx: Any, *args: Any) -> Any:
         i, w = ctx.saved_tensors
         assert i.requires_grad
         assert w.requires_grad
+        # We use ``detach()'' to ensure the backward call below does not
```
it really helps, I never realized that custom backward functions could themselves build a new autograd graph, I always assumed (I was wrong) that autograd was hands off there.. thanks for the comments @min-xu-ai !

blefaudeux reviewed (approved):
Looks great to me, feels like some of that could be sped up either with the recent cuda graph optimization from pytorch or by lowering this into triton or cuda kernels, but step by step is a good approach I think !

min-xu-ai commented:
> Looks great to me, feels like some of that could be sped up either with the recent cuda graph optimization from pytorch or by lowering this into triton or cuda kernels, but step by step is a good approach I think !

Couldn’t agree more!
